File: day1/day1.py
from collections import Counter

with open("day1.txt") as f:
    lists = f.read().splitlines()
    
first_col, second_col = [], []

for line in lists:
    first, second = line.split("   ")
    first = int(first)
    second = int(second)
    first_col.append(first)
    second_col.append(second)

first_col.sort()
second_col.sort()

total_dist = 0

for first, second in zip(first_col, second_col):
    total_dist += abs(first - second)

print(f"Part 1: {total_dist}")

# Part 2
part_two_total = 0

second_col_collection = Counter(second_col)
for item in first_col:
    part_two_total += item * second_col_collection[item]

print(f"Part 2: {part_two_total}")

File: day1/day1.rs
use std::iter::zip;

fn part1(list1: &Vec<i32>, list2: &Vec<i32>) -> i32 {
    let sum = zip(list1, list2)
        .map(|(a,b)| (a-b).abs())
        .sum::<i32>();

    sum
}

fn part2(list1: &Vec<i32>, list2: &Vec<i32>) -> i32 {
    list1
        .iter()
        .map(|&e| {
            let count = list2.iter().filter(|&&x| x == e).count() as i32;
            count * e
        })
        .sum::<i32>()
}


fn main() {
    let input = std::fs::read_to_string("day1.txt").unwrap();

    let mut list1 = vec![];
    let mut list2 = vec![];

    input
        .split("\n")
        .filter(|e| e.len() > 0)
        .map(|e| {
            let mut splits = e.split_whitespace();
            (splits.next().unwrap().parse::<i32>().unwrap(), splits.next().unwrap().parse::<i32>().unwrap())
        })
        .for_each(|(i,j)| {
            list1.push(i);
            list2.push(j);
        });

    list1.sort();
    list2.sort();

    println!("Result Part 1 = {}", part1(&list1, &list2));
    println!("Result Part 2 = {}", part2(&list1, &list2));
}